[TensorExpr] Resubmit: Introduce ExternalCall nodes to TE IR.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51594 [TensorExpr] Resubmit: Introduce ExternalCall nodes to TE IR.**

ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

The reason the PR was previously reverted was that the LLVM generated
calls to bridge functions were breaking unwind tables. This is now fixed
by requiring bridge functions to never throw and setting the
corresponding attribute in the LLVM generated code.

Differential Revision: [D26213882](https://our.internmc.facebook.com/intern/diff/D26213882)

Commit:
[TensorExpr] Resubmit: Introduce ExternalCall nodes to TE IR.

ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Resubmit: Introduce ExternalCall nodes to TE IR."


ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Resubmit: Introduce ExternalCall nodes to TE IR."


ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

The reason the PR was previously reverted was that the LLVM generated
calls to bridge functions were breaking unwind tables. This is now fixed
by requiring bridge functions to never throw and setting the
corresponding attribute in the LLVM generated code.

[ghstack-poisoned]

asuhan commented on torch/csrc/jit/tensorexpr/external_functions.cpp:
```diff
@@ -0,0 +1,123 @@
+  int64_t buf_dims_idx = 0;
+  for (int64_t i = 0; i < bufs_num; i++) {
+    buf_data_vec.push_back(buf_data[i]);
+    buf_dims_vec.push_back({});
```
I think the `clang-tidy` warning is worth fixing here.

asuhan commented on torch/csrc/jit/tensorexpr/external_functions.cpp:
```diff
@@ -0,0 +1,123 @@
+  } else {
+    try {
+      r = at::native::conv2d(x, w);
+    } catch (...) {
```
What's the plan for propagating error codes? I assume we could use `extra_args`, technically, but I don't like that solution.

ZolotukhinM replied:
We could change the signature to return an exit code, but the problem is that nothing else in our system expects to deal with such exit codes and knows how to raise them further. While it is possible to build all the infra for this, I think it's more practical to just check sizes/attributes of a node before we lower to NNC making sure a situation an exception occurs would never happen in the TE code.

asuhan reviewed (approved):
Good for quick re-landing, but let's have a follow-up soon bringing back error handling.

Commit:
clang-tidy on "[TensorExpr] Resubmit: Introduce ExternalCall nodes to TE IR."


ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

The reason the PR was previously reverted was that the LLVM generated
calls to bridge functions were breaking unwind tables. This is now fixed
by requiring bridge functions to never throw and setting the
corresponding attribute in the LLVM generated code.

Differential Revision: [D26213882](https://our.internmc.facebook.com/intern/diff/D26213882)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Resubmit: Introduce ExternalCall nodes to TE IR."


ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

The reason the PR was previously reverted was that the LLVM generated
calls to bridge functions were breaking unwind tables. This is now fixed
by requiring bridge functions to never throw and setting the
corresponding attribute in the LLVM generated code.

Differential Revision: [D26213882](https://our.internmc.facebook.com/intern/diff/D26213882)

[ghstack-poisoned]

Commit:
rebase on "[TensorExpr] Resubmit: Introduce ExternalCall nodes to TE IR."


ExternalCall nodes represent opaque calls to external functions to fill a
tensor (buffer) with values. It could be used to include nodes that are
otherwise not-representable as TE, or whose TE representation is currently too
slow.

To make an external function available in NNC as ExternalCall, one needs to
implement a "bridge" function that would take raw (void*) pointers to the data
along with the arrays containing dimension info. This function would then
internally call the desired external function and make sure the results of the
call are correctly placed in the provided raw data buffers.

The reason the PR was previously reverted was that the LLVM generated
calls to bridge functions were breaking unwind tables. This is now fixed
by requiring bridge functions to never throw and setting the
corresponding attribute in the LLVM generated code.

Differential Revision: [D26213882](https://our.internmc.facebook.com/intern/diff/D26213882)

[ghstack-poisoned]
